Add default "resolvedIf" function for tasks

binokaryg opened the pull request:
# Description

Adds a default resolvedIf function for tasks.

Issue medic/medic-conf#107

Documentation PR: medic/cht-docs#466

## defaultResolvedIf
- Uses [Utils.isFormSubmittedInWindow](https://github.com/medic/medic-nootils/blob/4339b25201bfbf51eef9179c8b7e0e92734a0d48/src/nootils.js#L57) to check if the resolving form has been submitted within a time period:
    - For contact-based tasks, the period is the same as the task window period
    - For report based tasks, the period is determined as follows:
        - start: start of the task window or 1 millisecond after reported date of the source report, whichever comes later
        - end: end of the task window
    - The resolving form is the task's first action form with 'report' type


## Usage
Options:
A. Skip `resolvedIf` section in the task definition, `defaultResolvedIf` will be used automatically
B. Using `this.defaultResolvedIf`:
1. Without specifying the parameters:
    ```
    resolvedIf: this.defaultResolvedIf,
    ```

2. Specifying parameters and optionally adding other conditions:
    ```
    resolvedIf: function (contact, report, event, dueDate) {
      return this.defaultResolvedIf(contact, report, event, dueDate) && ...
    },
    ```
3. Specifying the resolving form type if it is different from the default (form type specified in the first action of the task):
    ```
    resolvedIf: function (contact, report, event, dueDate) {
      return this.defaultResolvedIf(contact, report, event, dueDate, 'delivery');
    },
    ```

# Code review items

- [x] Readable: Concise, well named, follows the [style guide](https://docs.communityhealthtoolkit.org/contribute/code/style-guide/), documented if necessary.
- [x] Documented: Configuration and user documentation on [cht-docs](https://github.com/medic/cht-docs/)
- [x] Tested: Unit and/or integration tests where appropriate
- [x] Backwards compatible: Works with existing data and configuration. Any breaking changes documented in the release notes

.

# License

The software is provided under AGPL-3.0. Contributions to this project are accepted under the same license.

craig-landry commented:
@latin-panda would you please review this PR with @binokaryg?

latin-panda commented on src/nools/task-emitter.js:
```diff
@@ -163,4 +170,22 @@ function emitTasks(taskDefinition, Utils, Task, emit, c, r) {
 }
 
+function defaultResolvedIf (c, r, event, dueDate, resolvingForm, Utils) {
+  var start = 0;
```
Please use `cons` and `let` instead of `var`

Also here you can extract the resolvingForm and add validations. Remember to exit early to keep code flat and avoid unnecessary processing :)

latin-panda commented on src/nools/task-emitter.js:
```diff
@@ -6,7 +6,7 @@ function taskEmitter(taskDefinitions, c, Utils, Task, emit) {
   for (var idx1 = 0; idx1 < taskDefinitions.length; ++idx1) {
     taskDefinition = taskDefinitions[idx1];
-    prepareDefinition(taskDefinition);
+    prepareDefinition(taskDefinition, defaultResolvedIf, Utils);
```
We need to expose `defaultResolvedIf` in the task, right?

I think it comes more natural if defaults for tasks has a similar shape as a task definition and then we can concatenate both objects before passing it to `prepareDefinition()`, what do you think?

So, for example, let's say we have a file containing the definitions of defaults that we can export:
```
// File defining defaults

// Exporting obj similar as task
const taskDefaults = { 
   defaultResolvedIf: function(.....) {
      // do stuff;
   },
   
   // Here we can safely deprecate with backward compatibility or add more defaults in the future since objects can escalate easier.

}

exports {
   taskDefaults
}
```

Then we import `task defaults` and concatenate both objects:
```
// Task emitter file
 taskDefinition = { ...taskDefinitions[idx1], ...taskDefaults }; // <-- test if this works fine.
 prepareDefinition(taskDefinition);  // <-- it didn't changed
```

Unit testing will be more isolated because we just import the file for task defaults and test what it does (single responsibility), therefore the complexity of testing the task-emitter is reduced.

binokaryg replied:
Thanks for this great suggestion. It makes things more manageable. So the users will have to use `this.definition.defaultResolvedIf` instead of `this.defaultResolvedIf`? Should not be a problem if we document it nicely. Thanks.

latin-panda replied:
`this.definition.defaultResolvedIf`
I think it explains better where fn and props are coming from and keeps code simple and easy to maintain, but usability is important too. 
I'll leave it up to you on what works better for the end user :) Thank you for following up!

latin-panda commented on src/nools/task-emitter.js:
```diff
@@ -120,10 +120,17 @@ function emitTasks(taskDefinition, Utils, Task, emit, c, r) {
         readyStart: event.start || 0,
         readyEnd: event.end || 0,
         title: taskDefinition.title,
-        resolved: taskDefinition.resolvedIf(c, r, event, dueDate, scheduledTaskIdx),
```
If we do the task defaults file (from above comment) then that means taskDefinition already has the `defaultResolvedIf()` then you can use it from there: `taskDefinition.defaultResolvedIf(....)`

We can make a function that will handle that check. Just to give you an example:
```
callResolvedIf(.......) {
   if (.....) {
       return taskDefinition.resolvedIf(...);
   }

  return taskDefinition.defaultResolvedIf(...);
}

// then in the task def
task = {
   ... other props ...

   resolved: callResolvedIf(....),  
};
```

Because `emitForEvents` function is growing a lot and doing too many things, it should just emit and not do validations or other logic, but we'll fix it little by little as we go (not in this PR), so that's why I was thinking of reducing the default fn logic in this place and take it somewhere out. 
What do you think?

binokaryg replied:
I think this will work great. Thanks.

latin-panda commented on test/nools/mocks.js:
```diff
@@ -27,7 +27,7 @@ function aTask(type) {
     name: `task-${idCounter}`,
     title: [ { locale:'en', content:`Task ${idCounter}` } ],
-    actions: [ { form:'example-form' } ],
+    actions: [ {type: 'report', form:'example-form' } ],
```
The `type` isn't mandatory, so I think it would be good to keep a mock with out type and another mock with type defined to cover both scenarios.

latin-panda commented on test/nools/task-emitter.spec.js:
```diff
@@ -287,6 +288,136 @@ describe('task-emitter', () => {
         });
       });
 
+      it('resolvedIf is not required', () => {
```
When testing a common unit of logic, this case defaultResolveIf, it good to group it in a describe, it helps to use scoped before and and after hooks.
Sample:

```
describe('defaultResolveIf', () => {
  it('should  ___ when ____', () => {
      ....
  });

  it('should not  ___ when ____', () => {
      ....
  });
});
``

latin-panda reviewed (changes requested):
Nice job!! Thank you for picking up the task! 
I've left some comments inline, please let me know any question 👍

kennsippell commented on src/nools/definition-preparation.js:
```diff
@@ -4,10 +4,24 @@ Definition-preparation.js binds a value for `this` in all the functions within a
+        }
+        return defaultResolvedIf(contact, report, event, dueDate, reportAction.form, Utils);
+      }
+      return defaultResolvedIf(contact, report, event, dueDate, resolvingForm, Utils);
```
or if you define the defaultResolvedIf differently (perhaps via a [closure](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Closures), you won't need to pass the Utils lib when calling.  (same for resolvingForm?)

binokaryg replied:
Looks like we don't need to pass `Utils`  when running from CHT.  The tasks (tested only a few) are getting compiled and emitted successfully even when removing `Utils` from the [`taskEmitter`](https://github.com/medic/medic-conf/blob/ced75134d3ba38828d4ffef531eb2ef53a4db334/src/nools/lib.js#L10) and [`task-emitter.js`](https://github.com/medic/medic-conf/blob/ced75134d3ba38828d4ffef531eb2ef53a4db334/src/nools/task-emitter.js#L3)
However, if the `Utils` is not passed, the `medic-conf` tests fail with `ReferenceError: Utils is not defined`. Are we keeping Utils in task-emitter there only for the tests?

How do we solve this?

latin-panda replied:
Let me know if you still need help, we can look at it together in your code :)

binokaryg replied:
Fixed.

binokaryg commented on src/nools/definition-preparation.js:
```diff
@@ -4,10 +4,24 @@ Definition-preparation.js binds a value for `this` in all the functions within a
 
-function prepare(definition) {
+function prepare(definition, defaultResolvedIf, Utils) {
   var targetContext = {};
```
I had tried using `let` here at first but got this error from eslint: 
```
  8:7  error  Parsing error: Unexpected token targetContext
```
This is probably because of the `src/nools/.eslintrc` which has:

```
"parserOptions": {
    "ecmaVersion": 5
  },
```
Should we be changing it?

kennsippell replied:
medic-conf can run modern javascript, but the code in nools runs client-side and needs tighter restrictions.

I forget why it is ecmaVersion 5 though and not `2017` like we use in modern config. We need to cater to lowest common denominator...

I'm not sure if these conversations are still relevant. May break compatibility for very very old android versions (?) or cht-core v2.x users (?) I doubt it matters, but it's a change worth signalling if you go that path.
https://github.com/medic/medic-conf/issues/109#issuecomment-433002346

latin-panda replied:
Right, we need to test the app in crosswalk for oldie Android. 
We can try `"ecmaVersion": 6` which introduces `let, cons and arrow functions`, then make a medic-android apk with xwalk build it and run it. But I think it can work fine.

binokaryg replied:
I suggest that upgrading the `ecmaVersion` be part of another issue and PR.

latin-panda replied:
👍

binokaryg commented on src/nools/task-emitter.js:
```diff
@@ -120,10 +120,17 @@ function emitTasks(taskDefinition, Utils, Task, emit, c, r) {
+        task.resolved = taskDefinition.resolvedIf(c, r, event, dueDate, scheduledTaskIdx);
+      }
+      else {
+        var resolvingForm = taskDefinition.actions.find(function (action) { return action.type === 'report'; }).form;
```
> I think you can leave it as `defaultResolvedIf` responsibility of how to extract the form.
So we would have to pass the `taskDefinition` as a parameter to the `defaultResolvedIf`, right?

> Also the `find' function can return undefined
It is actually checked in `validate-declarative-schema.js` that if `resolvedIf` is not given, there should be at least one such report. Might not hurt to check it in case the validate logic changes. Thanks.

latin-panda replied:
> So we would have to pass the taskDefinition as a parameter to the defaultResolvedIf, right?

I think you can pass the `taskDefinition.actions` instead... unless you need more from taskDefinition.

> It is checked in validate-declarative-schema.js 

Awesome! Please add a unit test to address the scenario if possible :)

binokaryg replied:
As the `actions[n].type` is optional, I am thinking of changing it to:
```js
function getDefaultResolvingForm(taskDefinition) {
  var resolvingAction = taskDefinition && taskDefinition.actions &&
    taskDefinition.actions.find(
      function (action) {
        return action.type === 'report' || typeof action.type === 'undefined';
      }
    );
  return resolvingAction && resolvingAction.form;
}
```
Any suggestions?

latin-panda replied:
I think you can also check it as falsy value, it can compact to:
```
taskDefinition.actions.find(action => action.type === 'report' || !action.type);
```
if action.type is missing, does it get a default value or it's 'undefined' ?

binokaryg replied:
It is `undefined` at this point but I take your suggestion. Thanks!

binokaryg commented on src/nools/task-emitter.js:
```diff
@@ -163,4 +170,22 @@ function emitTasks(taskDefinition, Utils, Task, emit, c, r) {
+function defaultResolvedIf (c, r, event, dueDate, resolvingForm, Utils) {
+  var start = 0;
+  if (r) {//Report based task
+    //Start of the task window or after the report's reported date, whichever comes later
```
Thanks for the rephrase, it is clearer. Though it's not a day after but a millisecond after the reported date. So can I keep `just after the reported date` ?

latin-panda replied:
Perfect 👍

binokaryg commented on src/lib/validate-declarative-schema.js:
```diff
@@ -104,8 +104,12 @@ const TaskSchema = joi.array().items(
-    resolvedIf: joi.function().required()
-      .error(taskError('"resolvedIf" should be of type function(contact, report)')),
+    resolvedIf: joi.alternatives().conditional('actions', {
+      is: joi.array().items(joi.object({ type: 'contact' }).unknown()),
```
Thanks for this perspective. Did you mean something like this?
```js
    resolvedIf: joi
      .alternatives()
      .conditional('actions', {
        is: joi.array().has(joi.object({ type: 'report' }).unknown()),
        then: joi.function().optional(),
        otherwise: joi.function().required()
      })
```

latin-panda replied:
Looks good! :)

binokaryg commented on src/lib/validate-declarative-schema.js:
```diff
@@ -104,8 +104,12 @@ const TaskSchema = joi.array().items(
+      then: joi.function().required(),
+      otherwise: joi.function().optional()
+    })
+      .error(taskError('"resolvedIf" should be of type function(contact, report) if there is no report action')),
```
There can be multiple actions and it is sufficient to have one action with `type: report`. How do we convey it or we don't need to write everything in the error message?

latin-panda replied:
I'm not sure if I understand. You mean that you're getting this error for actions that are not 'report' ?
If that's the case, can you try [.when()](https://joi.dev/api/?v=17.4.0#anywhencondition-options)?

binokaryg replied:
I was talking about how we phrase the error message. My original error message was:
> ... if there is no report action

Your suggestion:
> ... if action type is not "report"

Your message is clearer but it does not say that there can be multiple actions and we only need one of them to be of report type.

```js
var resolvingForm = taskDefinition.actions.find(function (action) { return action.type === 'report'; }).form
```

latin-panda replied:
Oh I see, well "resolvedIf" should always be a function, so its type will be the same. 
What's different is if "resolvedIf" is mandatory or not when defining each action.

> Schema error in actions array: Actions with property "type" which value is different than "report", should define property "resolvedIf" as: function(contact, report) { ... }.

This is very specific, what do you think?  🤔

binokaryg replied:
Yes, it is good, but we might also have to consider that `type` is also optional.

binokaryg replied:
#### `resolvedIf` is optional in these cases:
1. One action with type: `report`
```js
actions: [
            {
                type: 'report',
                form: 'pregnancy_followup',
                label: 'Pregnancy followup'
            }
        ],
```

2. Two actions: one is `contact` type, another is `report`
```js
actions: [
            {
                type: 'contact',
            },
            {
                type: 'report',
                form: 'pregnancy_followup',
                label: 'Pregnancy followup'
            }
        ],
```
3. One action, type is not defined, defaults to report
```js
actions: [
            {
                form: 'pregnancy_followup',
                label: 'Pregnancy followup'
            }
        ],
```

#### `resolvedIf` is mandatory in this case:
1. One action with type: 'contact'
```js
actions: [
            {
                type: 'contact',
                label: 'New contact'
            }
        ],
```

latin-panda replied:
Nice! This last comment is good and nice to add in the ticket, it'll help QA when testing the feature. Also for any training doc out there. 

Regarding the error message, it still applies since when the `action.type` is missing then it's defaulting to a value.

However, if historically this leads to a frequent struggle, then we can warn the user that an `action.type` is going to default to a value, when processing the tasks (before upload of settings). So it's a different message, for example:

`WARN: Target's action is missing the property type, will default to [something]`

`ERROR: Schema error in actions array: Actions with property "type" which value is different than "report", should define property "resolvedIf" as: function(contact, report) { ... }.`

What do you think? :)

binokaryg replied:
Sounds good, where do you suggest putting the warning?

latin-panda replied:
I wonder if we can do something in the Joi validation using message() and warning() or perhaps using a combination with when().then() 🤔 Otherwise maybe just after the Joi schema validation, we can create a function that checks and print the warning.

binokaryg commented on test/run-nools-lib.js:
```diff
@@ -19,6 +19,20 @@ const runNoolsLib = ({ c, targets, tasks }) => {
       },
       now: () => new Date(TEST_DATE),
       isTimely: function() { return true; },
+      isFormSubmittedInWindow: function(reports, form, start, end, count) {
```
I am also having the problem of `Utils` not being accessible inside functions such as `defaultResolvedIf` in `task-defaults.js` when running the unit tests. It runs fine in the app though as `Utils` is [added ](https://github.com/medic/cht-core/blob/0d4c275b088a92edb5b971d4f40aafcde3f93d42/shared-libs/rules-engine/src/rules-emitter.js#L49)to the scope when compiling `nools`.

latin-panda replied:
If you still need help with this one, let me know and we can see it together :)

binokaryg replied:
My problem has been fixed with help from @kennsippell  but your concern regarding the possible deviation from the original code remains.

latin-panda replied:
Some ideas, but test if they are feasible :) 

- If it doesn't matter to test isFormSubmittedInWindow content, then we can make a mock object and stub a function by using sinon, send this object to runNoolsLib as a Utils to use. Sample:
``` 
** Your test file **

utilsMock = {
  isFormSubmittedInWindow: sinon.stub().returns(true) // or false depending on your test
};
....
const { emitted } = runNoolsLib({c, tasks, targets, utilsMock}); // pass it to the context

expect(utilsMock.isFormSubmittedInWindow.callCount).to.equal(1); // Now we can check this :) 

** runNoolsLib file **

const runNoolsLib = ({ c, targets, tasks, utilsMock }) => {
  const context = {
     ......
     Utils: {...utilsMock}  // check if utilsMock is defined then use it, otherwise use the existent code here. So we don't break other tests. But from new dev & forward use the utilsMock.
      ....
  }
```

- if we care about testing isFormSubmittedInWindow content, then it's better to use the original code. If possible import the original Utils. Sample:
```
// runNoolsLib file
const Utils = require(....) // 

const runNoolsLib = ({ c, targets, tasks}) => {
  const context = {
     ......
     Utils: {
          ....
          isFormSubmittedInWindow: Utils.isFormSubmittedInWindow,
          ...
      }
      ....
  }
```

I usually do the first one when the dependency lib, this case Utils.isFormSubmittedInWindow, has already good unit test coverage. It's verified that piece is already working.

Share with me if you find any other alternative to learn as well :)

binokaryg replied:
>  Utils: {...utilsMock}
We also use `Utils.addDate` in `defaultResolvedIf`. Shall we add it too to the `utilsMock`?

latin-panda replied:
yes, I think so for more accurate testing, you can add as needed in your testing.

binokaryg commented on src/lib/validate-declarative-schema.js:
```diff
@@ -104,8 +104,12 @@ const TaskSchema = joi.array().items(
       .error(taskError('"contactLabel" should either be a non-empty string or function(contact, report)')),
-    resolvedIf: joi.function().required()
-      .error(taskError('"resolvedIf" should be of type function(contact, report)')),
+    resolvedIf: joi.alternatives().conditional('actions', {
```
Thanks, I'm changing the format for `resolvedIf` only for now.

latin-panda commented on src/nools/task-defaults.js:
```diff
@@ -0,0 +1,38 @@
+}
+
+function getDefaultResolvingForm(taskDefinition) {
+  var resolvingAction = taskDefinition && taskDefinition.actions &&
```
You can exit early, it'll require less effort to understand the conditions and will keep lines shorter. Specially if we're calling a function with a callback, in that case better assign the returning value to a variable and then use the variable in the condition.
Usually we treat falsy values ("", null, undefined, false, 0) the same, so we can avoid using "typeof" which is for specific cases where we need to differentiate from any of those values, for example. (Is it a specific case?)

```
function getDefaultResolvingForm(taskDefinition) {
  var resolvingAction;    // --> because of using "var", better here to avoid hoisting

  if (!taskDefinition || !taskDefinition.actions) { // --> exit early
    return;
  }

  resolvingAction = taskDefinition.actions.find(
      function (action) {
        return !action.type || action.type === 'report' // --> no typeof
      }
    );

  return resolvingAction && resolvingAction.form;
}
```

latin-panda commented on test/nools/task-emitter.spec.js:
```diff
@@ -657,6 +670,72 @@ describe('task-emitter', () => {
+        delete config.tasks[0].resolvedIf;
+        // when
+        const { emitted } = runNoolsLib(config);
+        //then
```
```suggestion
        
        // then
```

latin-panda commented on test/nools/task-emitter.spec.js:
```diff
@@ -657,6 +670,72 @@ describe('task-emitter', () => {
+          resolved: false,//not resolved
+        });
+
+        //given
```
Optionally from here and down can be a seperate test case "it()"

latin-panda commented on test/run-nools-lib.js:
```diff
@@ -3,31 +3,36 @@ const taskEmitter = require('../src/nools/task-emitter');
+      if (utilsMock) {
+        return utilsMock;
+      }
+      return {
```
You can add a comment or (ToDo comment) explaining that the prefer way is to use 'utilsMocks' and this one will be removed when the unit test are alined with the new approach.
See comment below

latin-panda commented on test/run-nools-lib.js:
```diff
@@ -42,8 +47,8 @@ const runNoolsLib = ({ c, targets, tasks }) => {
 
-  targetEmitter(context.targets, context.c, context.Utils, context.Target, context.emit);
-  taskEmitter(context.tasks, context.c, context.Utils, context.Task, context.emit);
+  targetEmitter(context.targets, context.c, context.Utils(), context.Target, context.emit);
```
```suggestion
  targetEmitter(context.targets, context.c, utilsMock || context.Utils, context.Target, context.emit);
```
I think we can keep the old Utils as object with a comment saying that it's deprecated and eventually will be removed, so other devs don't add more logic to that, and then make the conditional here.

latin-panda commented on test/run-nools-lib.js:
```diff
@@ -42,8 +47,8 @@ const runNoolsLib = ({ c, targets, tasks }) => {
-  targetEmitter(context.targets, context.c, context.Utils, context.Target, context.emit);
-  taskEmitter(context.tasks, context.c, context.Utils, context.Task, context.emit);
+  targetEmitter(context.targets, context.c, context.Utils(), context.Target, context.emit);
+  taskEmitter(context.tasks, context.c, context.Utils(), context.Task, context.emit);
```
```suggestion
  // context.Utils is deprecated, please use utilsMock
  taskEmitter(context.tasks, context.c, utilsMock || context.Utils, context.Task, context.emit);
```

latin-panda reviewed (changes requested):
Nice job!! Turning out great 👍! I left some minor comments/questions in line see above and below 

Can you please add the steps to test this in the ticket (as you did in the PR) recently we added this step in [the workflow of CHT](https://docs.communityhealthtoolkit.org/contribute/code/workflow/#updating-the-issue-with-what-you-actually-did) as easier way for QA, PO, Designers to understand the solution design for a ticket.

Thank you!

binokaryg commented on src/nools/task-defaults.js:
```diff
@@ -0,0 +1,38 @@
+      if (typeof resolvingForm !== 'string') {
+        throw new Error('Could not find the default resolving form!');
+      }
+      var start = 0;
```
Thanks! Moved the declarations to the top.

binokaryg commented:
> Nice job!! Turning out great 👍! I left some minor comments/questions in line see above and below
> 
> Can you please add the steps to test this in the ticket (as you did in the PR) recently we added this step in [the workflow of CHT](https://docs.communityhealthtoolkit.org/contribute/code/workflow/#updating-the-issue-with-what-you-actually-did) as easier way for QA, PO, Designers to understand the solution design for a ticket.
> 
> Thank you!

Thanks @latin-panda, I have added the changes that you had suggested. Thank you again for all the great feedback.

latin-panda commented on src/nools/task-defaults.js:
```diff
@@ -0,0 +1,49 @@
+// File defining task defaults
+function taskDefaults(Utils) {
```
Nice! Another alternative is to pass the `Utils` as parameter to `defaultResolvedIf`, then the module can return the object directly and avoid higher level variables. But then `defaultResolvedIf` would need this extra parameter every-time we call it.
Both have pros & cons and both ways work :) 
```
/// task-defaults.js
module.exports = {
{
    defaultResolvedIf: function (contact, report, event, dueDate, Utils) { 
          ...... 
    }
}

/// task-emitter.js

if (typeof taskDefinition.resolvedIf !== 'function') {
      taskDefinition.resolvedIf = function (contact, report, event, dueDate) { taskDefinition.defaultResolvedIf(contact, report, event, dueDate, Utils) }; 
     // new Javascript would be: taskDefinition.resolvedIf = (...args) => taskDefinition.defaultResolvedIf(...args, Utils)
}
```

latin-panda replied:
✨✨ best of both worlds!✨✨

latin-panda commented on test/run-nools-lib.js:
```diff
@@ -3,31 +3,37 @@ const taskEmitter = require('../src/nools/task-emitter');
-      },
-      now: () => new Date(TEST_DATE),
-      isTimely: function() { return true; },
+    Utils: () => {
```
Let me know your thoughts in this one~

binokaryg replied:
Yes, that was an oversight from me, thanks for pointing it out. :)

latin-panda reviewed (approved):
✅ Great job! It looks good to me and I think is ready for AT.
Thank you for addressing all the suggestions and submitting this code 🙏
